File: app/src/main/java/com/krystianwsul/checkme/gui/friends/FindFriendActivity.java
package com.krystianwsul.checkme.gui.friends;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.ValueEventListener;
import com.krystianwsul.checkme.MyCrashlytics;
import com.krystianwsul.checkme.R;
import com.krystianwsul.checkme.firebase.DatabaseWrapper;
import com.krystianwsul.checkme.firebase.UserData;
import com.krystianwsul.checkme.gui.MainActivity;

import junit.framework.Assert;

public class FindFriendActivity extends AppCompatActivity {
    private static final String USER_KEY = "user";
    private static final String LOADING_KEY = "loading";

    private EditText mFindFriendEmail;

    private LinearLayout mFindFriendUserLayout;
    private TextView mFindFriendUserName;
    private TextView mFindFriendUserEmail;

    private LinearLayout mFindFriendProgress;

    private boolean mLoading = false;
    private UserData mUserData = null;

    private DatabaseReference mDatabaseReference;
    private ValueEventListener mValueEventListener;

    @NonNull
    public static Intent newIntent(@NonNull Context context) {
        return new Intent(context, FindFriendActivity.class);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_find_friend, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_search:
                startSearch();

                break;
            default:
                throw new UnsupportedOperationException();
        }
        return true;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_find_friend);

        Log.e("asdf", "onCreate " + hashCode());

        Toolbar toolbar = (Toolbar) findViewById(R.id.find_friend_toolbar);
        Assert.assertTrue(toolbar != null);

        setSupportActionBar(toolbar);

        mFindFriendEmail = (EditText) findViewById(R.id.find_friend_email);
        Assert.assertTrue(mFindFriendEmail != null);

        mFindFriendEmail.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_SEARCH) {
                startSearch();
                return true;
            }
            return false;
        });

        mFindFriendUserLayout = (LinearLayout) findViewById(R.id.find_friend_user_layout);
        Assert.assertTrue(mFindFriendUserLayout != null);

        mFindFriendUserLayout.setOnClickListener(v -> {
            Assert.assertTrue(mUserData != null);
            Assert.assertTrue(!mLoading);

            UserData myUserData = MainActivity.getUserData();
            Assert.assertTrue(myUserData != null);

            DatabaseWrapper.addFriend(myUserData, mUserData);

            finish();
        });

        mFindFriendUserName = (TextView) findViewById(R.id.find_friend_user_name);
        Assert.assertTrue(mFindFriendUserName != null);

        mFindFriendUserEmail = (TextView) findViewById(R.id.find_friend_user_email);
        Assert.assertTrue(mFindFriendUserEmail != null);

        mFindFriendProgress = (LinearLayout) findViewById(R.id.find_friend_progress);
        Assert.assertTrue(mFindFriendProgress != null);
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);

        if (savedInstanceState != null) {
            Assert.assertTrue(savedInstanceState.containsKey(LOADING_KEY));
            mLoading = savedInstanceState.getBoolean(LOADING_KEY);

            if (savedInstanceState.containsKey(USER_KEY)) {
                mUserData = savedInstanceState.getParcelable(USER_KEY);
                Assert.assertTrue(mUserData != null);
            }
        }

        updateLayout();
    }

    @Override
    protected void onStart() {
        super.onStart();

        if (mLoading)
            loadUser();
    }

    private void loadUser() {
        Assert.assertTrue(mLoading);
        Assert.assertTrue(mUserData == null);
        Assert.assertTrue(!TextUtils.isEmpty(mFindFriendEmail.getText()));

        String key = UserData.getKey(mFindFriendEmail.getText().toString());

        Log.e("asdf", "starting");

        mValueEventListener = new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                Log.e("asdf", "onDataChange " + hashCode());

                Assert.assertTrue(dataSnapshot != null);

                mDatabaseReference.removeEventListener(mValueEventListener);

                mLoading = false;
                mValueEventListener = null;
                mDatabaseReference = null;

                if (dataSnapshot.exists()) {
                    mUserData = dataSnapshot.getValue(UserData.class);
                    Assert.assertTrue(mUserData != null);
                } else {
                    Toast.makeText(FindFriendActivity.this, R.string.userNotFound, Toast.LENGTH_SHORT).show();
                }

                updateLayout();
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {
                Assert.assertTrue(databaseError != null);

                mDatabaseReference.removeEventListener(mValueEventListener);

                mLoading = false;
                mValueEventListener = null;
                mDatabaseReference = null;

                updateLayout();

                MyCrashlytics.logException(databaseError.toException());

                Log.e("asdf", "onCancelled", databaseError.toException());

                Toast.makeText(FindFriendActivity.this, R.string.connectionError, Toast.LENGTH_SHORT).show();
            }
        };

        mDatabaseReference = DatabaseWrapper.getUserDataDatabaseReference(key);

        Log.e("asdf", "addValueEventListener " + mValueEventListener.hashCode());
        mDatabaseReference.addValueEventListener(mValueEventListener);
    }

    private void updateLayout() {
        Log.e("asdf", "updateLayout " + hashCode());

        if (mUserData != null) {
            Assert.assertTrue(!mLoading);

            mFindFriendEmail.setEnabled(true);
            mFindFriendUserLayout.setVisibility(View.VISIBLE);
            mFindFriendProgress.setVisibility(View.GONE);

            mFindFriendUserName.setText(mUserData.getDisplayName());
            mFindFriendUserEmail.setText(mUserData.getEmail());
        } else if (mLoading) {
            mFindFriendEmail.setEnabled(false);
            mFindFriendUserLayout.setVisibility(View.GONE);
            mFindFriendProgress.setVisibility(View.VISIBLE);
        } else {
            mFindFriendEmail.setEnabled(true);
            mFindFriendUserLayout.setVisibility(View.GONE);
            mFindFriendProgress.setVisibility(View.GONE);
        }
    }

    @Override
    protected void onStop() {
        super.onStop();

        Log.e("asdf", "onStop");

        if (mLoading) {
            Assert.assertTrue(mDatabaseReference != null);
            Assert.assertTrue(mValueEventListener != null);

            Log.e("asdf", "removing listener " + mValueEventListener.hashCode());

            mDatabaseReference.removeEventListener(mValueEventListener);
        } else {
            Assert.assertTrue(mDatabaseReference == null);
            Assert.assertTrue(mValueEventListener == null);
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

        outState.putBoolean(LOADING_KEY, mLoading);

        if (mUserData != null)
            outState.putParcelable(USER_KEY, mUserData);
    }

    private void startSearch() {
        if (TextUtils.isEmpty(mFindFriendEmail.getText()))
            return;

        mLoading = true;
        mUserData = null;

        updateLayout();

        loadUser();
    }
}

File: app/src/main/java/com/krystianwsul/checkme/gui/friends/FindFriendActivity.kt
package com.krystianwsul.checkme.gui.friends

import android.Manifest
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.*
import android.view.inputmethod.EditorInfo
import androidx.activity.viewModels
import androidx.core.view.isInvisible
import androidx.core.view.isVisible
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.Snackbar
import com.jakewharton.rxbinding4.view.keys
import com.jakewharton.rxbinding4.widget.editorActionEvents
import com.krystianwsul.checkme.R
import com.krystianwsul.checkme.databinding.ActivityFindFriendBinding
import com.krystianwsul.checkme.databinding.RowListAvatarBinding
import com.krystianwsul.checkme.gui.base.NavBarActivity
import com.krystianwsul.checkme.utils.animateVisibility
import com.krystianwsul.checkme.utils.ignore
import com.krystianwsul.checkme.utils.loadPhoto
import com.krystianwsul.checkme.utils.toV3
import com.krystianwsul.checkme.viewmodels.FindFriendViewModel
import com.tbruyelle.rxpermissions2.RxPermissions
import io.reactivex.rxjava3.kotlin.addTo
import io.reactivex.rxjava3.kotlin.plusAssign

class FindFriendActivity : NavBarActivity() {

    companion object {

        fun newIntent(context: Context) = Intent(context, FindFriendActivity::class.java)
    }

    private val viewModel by viewModels<FindFriendViewModel>()

    override val rootView get() = binding.findFriendRoot

    private lateinit var binding: ActivityFindFriendBinding

    private val adapter by lazy { Adapter() }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_find_friend, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_search -> startSearch()
            else -> throw UnsupportedOperationException()
        }
        return true
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityFindFriendBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setSupportActionBar(binding.findFriendToolbar)

        binding.findFriendEmail.apply {
            createDisposable += editorActionEvents {
                if (it.actionId == EditorInfo.IME_ACTION_SEARCH) {
                    startSearch()
                    true
                } else {
                    false
                }
            }.subscribe()

            createDisposable += keys {
                if (it.action == KeyEvent.ACTION_DOWN && it.keyCode == KeyEvent.KEYCODE_ENTER) {
                    startSearch()
                    true
                } else {
                    false
                }
            }.subscribe()
        }

        binding.findFriendRecycler.let {
            it.layoutManager = LinearLayoutManager(this)
            it.adapter = adapter
        }

        viewModel.viewStateObservable
                .subscribe(::updateLayout)
                .addTo(createDisposable)
    }

    private fun updateLayout(state: FindFriendViewModel.ViewState) {
        val hide = mutableListOf<View>()
        val show = mutableListOf<View>()

        @Suppress("IMPLICIT_CAST_TO_ANY")
        when (state) {
            FindFriendViewModel.ViewState.Permissions ->
                RxPermissions(this).request(Manifest.permission.READ_CONTACTS)
                        .toV3()
                        .subscribe { viewModel.viewActionRelay.accept(FindFriendViewModel.ViewAction.Permissions(it)) }
                        .addTo(createDisposable)
            is FindFriendViewModel.ViewState.Loading -> {
                binding.findFriendEmail.isEnabled = false

                hide += binding.findFriendRecycler
                show += binding.findFriendProgress
            }
            is FindFriendViewModel.ViewState.Loaded -> {
                binding.findFriendEmail.isEnabled = true

                show += binding.findFriendRecycler
                hide += binding.findFriendProgress

                adapter.submitList(state.contacts) { binding.findFriendRecycler.smoothScrollToPosition(0) }
            }
            is FindFriendViewModel.ViewState.Error ->
                Snackbar.make(binding.root, state.stringRes, Snackbar.LENGTH_SHORT).show()
        }.ignore()

        animateVisibility(show, hide)
    }

    private fun startSearch() {
        binding.findFriendEmail
                .text
                .toString()
                .takeIf { it.isNotEmpty() }
                ?.let { viewModel.viewActionRelay.accept(FindFriendViewModel.ViewAction.Search(it)) }
    }

    private inner class Adapter : ListAdapter<FindFriendViewModel.Contact, Holder>(
            object : DiffUtil.ItemCallback<FindFriendViewModel.Contact>() {

                override fun areItemsTheSame(
                        oldItem: FindFriendViewModel.Contact,
                        newItem: FindFriendViewModel.Contact,
                ) = oldItem.email == newItem.email

                override fun areContentsTheSame(
                        oldItem: FindFriendViewModel.Contact,
                        newItem: FindFriendViewModel.Contact,
                ) = oldItem == newItem
            }
    ) {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
                Holder(RowListAvatarBinding.inflate(layoutInflater, parent, false))

        override fun onBindViewHolder(holder: Holder, position: Int) {
            holder.binding.apply {
                getItem(position).apply {
                    root.setOnClickListener {
                        viewModel.viewActionRelay.accept(FindFriendViewModel.ViewAction.AddFriend)
                        finish()
                    }

                    rowListAvatarImage.loadPhoto(photoUri)

                    rowListAvatarName.text = displayName
                    rowListAvatarDetails.text = email

                    rowListAvatarChildren.isVisible = false

                    rowListAvatarSeparator.isInvisible = position == itemCount - 1
                }
            }
        }
    }

    private class Holder(val binding: RowListAvatarBinding) : RecyclerView.ViewHolder(binding.root)
}
